modules/generator/registry/counter.go

```diff
@@ -0,0 +1,140 @@
+package registry
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"go.uber.org/atomic"
+)
+
+type counter struct {
+	name   string
+	labels []string
+
+	// seriesMtx is used to sync modifications to the map, not to the data in series
+	seriesMtx sync.RWMutex
+	series    map[uint64]*counterSeries
+
+	onAddSeries    func(count uint32) bool
```

Comment:
I think calling these callbacks can be a bit confusing since they intervine in the creation of the metrics (i.e. the result of `onAddSeries` will determine if the series is created or not).

Reply:
I agree these things are confusing. I don't like we have to use anonymous functions to couple the counters/histograms to the registry, but I'm not sure what would work better. I wanted to keep this functions somewhat generic so it's easier to test. Injecting the full `managedRegistry` means we have to test them together all the time.

Part of the complexity is that `onAddSeries` is used for both checking whether we can add series and to keep count of them. We could split this up into 2 callbacks: `canAddSeries` and `incActiveSeries` but this is 1) more work and 2) might introduce a race condition between asking whether you can add series and reporting you added them.